x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -508,6 +511,144 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+            NumberFieldMapper.NumberFieldType maxFieldType = metricFields.get(Metric.max);
+            NumberFieldMapper.NumberFieldType sumFieldType = metricFields.get(Metric.sum);
+            NumberFieldMapper.NumberFieldType countFieldType = metricFields.get(Metric.value_count);
+
+            private AggregateMetricDoubleBlockLoader() {}
+
+            static NumericDocValues getNumericDocValues(NumberFieldMapper.NumberFieldType field, LeafReader leafReader) throws IOException {
+                if (field == null) {
+                    return null;
+                }
+                String fieldName = field.name();
+                var values = leafReader.getNumericDocValues(fieldName);
+                if (values != null) {
+                    return values;
+                }
+
+                var sortedValues = leafReader.getSortedNumericDocValues(fieldName);
+                return DocValues.unwrapSingleton(sortedValues);
+            }
+
+            @Override
+            public AllReader reader(LeafReaderContext context) throws IOException {
+                NumericDocValues minValues = getNumericDocValues(minFieldType, context.reader());
+                NumericDocValues maxValues = getNumericDocValues(maxFieldType, context.reader());
+                NumericDocValues sumValues = getNumericDocValues(sumFieldType, context.reader());
+                NumericDocValues valueCountValues = getNumericDocValues(countFieldType, context.reader());
+
+                if (minValues == null || maxValues == null || sumValues == null || valueCountValues == null) {
+                    throw new UnsupportedOperationException("Must have all subfields to use aggregate double metric in ESQL");
+                }
+                return new BlockDocValuesReader() {
+
+                    private int docID = -1;
+
+                    @Override
+                    protected int docId() {
+                        return docID;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "BlockDocValuesReader.AggregateMetricDouble";
+                    }
+
+                    @Override
+                    public Block read(BlockFactory factory, Docs docs) throws IOException {
+                        try (var builder = factory.aggregateMetricDoubleBuilder(docs.count())) {
+                            copyDoubleValuesToBuilder(docs, builder.min(), minValues);
+                            copyDoubleValuesToBuilder(docs, builder.max(), maxValues);
+                            copyDoubleValuesToBuilder(docs, builder.sum(), sumValues);
+                            copyIntValuesToBuilder(docs, builder.count(), valueCountValues);
+                            return builder.build();
+                        }
+                    }
+
+                    private void copyDoubleValuesToBuilder(Docs docs, BlockLoader.DoubleBuilder builder, NumericDocValues values)
```

Comment:
:+1:

Reply:
I suppose you could reuse one of the other builders somehow maybe? Like the one that numerics uses for doubles. It has code just like this, right? Too tricky to share?